### DISCUSSION TITLE
How to make Record case insensitive.

### DISCUSSION THREAD

**Discussion by TruePluto** [dotnet/csharplang#6994] [General] [answered]
I have some code such as following
```cs
Person person1 = new(FirstName: "Nancy", LastName: "davolio"); //lowercase
Person person2 = new(FirstName: "Nancy", LastName: "Davolio");
Person person3 = new(FirstName: "Nancy", LastName: "Davolio");
Console.WriteLine(value: person1 == person2); // output:expect True but actually False ----
Console.WriteLine(value: person3 == person2); // output: True
Console.WriteLine(value: ReferenceEquals(objA: person1, objB: person2)); // output: False
Console.WriteLine(value: ReferenceEquals(objA: person3, objB: person2)); // output: False
Console.WriteLine(value: "Hello, World!");

public record Person(string FirstName, string LastName);
```
Output is 
```
False ----here-----
True
False
False
Hello, World!
```
Meaning `Record` is case sensitive， but I want to make it case insensitive. what am I missing?
Thanks for your attention.

**Comment by CyrusNajmabadi** [comment, marked as answer]
Override .Equals in your record to implement whatever equality semantics you want.

**Reply by TruePluto** [reply]
Could you give me some code as an example. thanks

**Reply by CyrusNajmabadi** [reply]
@TruePluto: `public override bool Equals(Person p) { ... }`. Put whatever logic you want in that to compare your instance to 'p'.

**Reply by TruePluto** [reply]
I made a console app following your advice
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```
and C#
```cs
Person p1 = new Person("HELLO", "KITTY");
Person p2 = new Person("hello", "kitty");
Console.WriteLine(p1==p2);  


public record Person(string FirstName, string LastName)
{
    public override bool Equals(Person? p)
    {
        if(p == null) return false; 
        if(p == this) return true;
        return
            string.Equals(FirstName, p.FirstName, StringComparison.InvariantCultureIgnoreCase) &&
            string.Equals(LastName, p.LastName, StringComparison.InvariantCultureIgnoreCase)
            ;
    }
}
```
I got a compile error 
`"CS0115 Person.Equals(Person?)" no suitable method found to override`

Thank you  for your attention

**Reply by yaakov-h** [reply]
https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ6AKBTMBnAeygAIAHJYgXmKmwHdid8iAKAIgAkBRAGR4Hk2AGmJsA0gEkAKlICabAJQBuLLkIlSCKjXqM1rNgAtsAGxMFhogNYBLYMACeilWjhIAnC3KVKm5cWJ0INcAZmIwbABjAjAAEz1mKBY3AAZiADEbfGAAOQBDAFtsEVTiHjy8XMLsBXQAb3QAgLgwgDcs4Ag8k2IAIwICHq4ARy6TPBYmdQB+Mlq0JuIGhcWmmwAzFgAlbHXcbChI7BGxidIRKAgzBQViOAB2YnXuvGwlQJXV4g3t3f3D46jF5eETAQw2PA3O6PYCQN6NL7QhGIpqpAB0J2BmWy+SKIlIaOxlVxxWIAGVYTYoABzADCBAKpDyYAhRDREigrWZNjyUGAtKunQiEmpUBi2FpFRqxAAZDLkSi7kgUhigeMWOVidV8WjNVU8eTKTT6Yzuep2Zzubz+YKIMLReLJa95oqmi5FgBfdAeoA

**Reply by DavidArno** [reply]
@yaakov-h, `warning CS8851: 'Person' defines 'Equals' but not 'GetHashCode'` is significant here. `GetHashCode` needs to be redefined to handle the case-insensitivity.

**Reply by yaakov-h** [reply]
only if you make use of the hashcode - fortunately there's a nice compiler warning there as a reminder

**Comment by TruePluto** [comment]
I want to use the class person as key for a dictionary class such as  
```cs
public class Portfolio{...}
public record Person (string FirstName, string LastName);
public class PersonProfolioDict : Dictionary<Person, Portfolio> {...};
```
So I believe the function `GetHashCode()` is necessary.  I wrote the code as following
```csharp
public record Person(string FirstName, string LastName)
{
    public virtual bool Equals(Person? p)
    {
        if (ReferenceEquals(p, null)) return false;
        if (ReferenceEquals(p, this)) return true;
        return
            string.Equals(FirstName, p.FirstName, StringComparison.InvariantCultureIgnoreCase) &&
            string.Equals(LastName, p.LastName, StringComparison.InvariantCultureIgnoreCase)
            ;
    }
    public override int GetHashCode()
    {
        return HashCode.Combine(FirstName.ToUpperInvariant(), LastName.ToUpperInvariant());
        //for Net <6.0
        //return FirstName.ToUpperInvariant( ).GetHashCode() ^ LastName.ToUpperInvariant().GetHashCode();
    }
}
```
Is it correct? 
I don't know whether it is a good idea to give attributes to indicate the string compare mode 
such as 
```cs
public record Person
    (
        [StringComparison(StringComparison.InvariantCultureIgnoreCase)]
        string FirstName,
        [StringComparison(StringComparison.InvariantCultureIgnoreCase)]
        string LastName
    );
```
Thank you all for your attention!

**Reply by yaakov-h** [reply]
`StringComparer.InvariantCultureIgnoreCase.GetHashCode(...)` should avoid a string allocation and therefore be a little bit faster.

**Reply by CyrusNajmabadi** [reply]
> I don't know whether it is a good idea to give attributes to indicate the string compare mode such as

It's a fine idea.  We leave it to Roslyn or the community to make a determination if someone wants to write a generator that would key off such things like that and generate appropriate Equals/GetHashCode overrides.  It's explicitly been chosen to be external to the language.  :)

Thanks!

**Comment by iam3yal** [comment]
Another alternative is to do something like this:
```csharp
public sealed record Person(Name FirstName, Name LastName);

public readonly struct Name : IEquatable<Name>
{
    public readonly string _value;

    public Name(string value)
       => _value = value;

    public bool Equals(Name other)
    {
        return string.Equals(
            _value, 
            other._value, 
            StringComparison.InvariantCultureIgnoreCase);
    }

    public override int GetHashCode()
        => StringComparer.InvariantCultureIgnoreCase.GetHashCode(_value);

    public static implicit operator string(Name name) => name._value;
    public static implicit operator Name(string value) => new Name(value);
}
```
